GetStream/stream-chat-android
# [2946] Handle system message from channel truncated event

https://github.com/GetStream/stream-chat-android/issues/2946

### 🎯 Goal

Android SDK doesn't respect the system message sent with the `channel.truncated` event. Instead, it relies on the `message.new` event that can arrive with a delay. The task is to make a transition between (a list of message) to (truncated channel with system message) as seamless as possible.

### 🛠 Implementation details

Persisting the system message coming with `channel.truncated` event instead of waiting for the `message.new` event to update the UI.

### 🎨 UI Changes

The recording on the left shows "No messages" placeholder blinking for a fraction of a second when the channel is truncated.

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150024690-eca932bb-2577-44d3-bc4c-f737d1c4c920.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150024705-8ae17b6b-c282-465b-aa97-e6c695fc69a1.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

1. Check out `bugfix/2946-handle-system-message-from-channel-truncated-event` branch and apply the patch below
2. Navigate to the message list screen and tap on the avatar in the message list header
3. Observe the truncated channel with a system message  

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt	(revision 851d0f068967850baf7f5e32a5902563011d122f)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt	(date 1642542167472)
@@ -4,11 +4,12 @@
 import androidx.lifecycle.MediatorLiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.models.Member
+import io.getstream.chat.android.client.models.Message
 import io.getstream.chat.android.livedata.ChatDomain
 import io.getstream.chat.android.livedata.controller.ChannelController
 import io.getstream.chat.android.livedata.utils.Event
-import io.getstream.chat.ui.sample.util.extensions.isAnonymousChannel
 
 class ChatViewModel(private val cid: String, private val chatDomain: ChatDomain = ChatDomain.instance()) : ViewModel() {
 
@@ -31,16 +32,10 @@
     fun onAction(action: Action) {
         when (action) {
             is Action.HeaderClicked -> {
-                val controller = requireNotNull(channelController)
-                _navigationEvent.value = Event(
-                    if (action.members.size > 2 ||
-                        controller.offlineChannelData.value?.isAnonymousChannel() == false
-                    ) {
-                        NavigationEvent.NavigateToGroupChatInfo(cid)
-                    } else {
-                        NavigationEvent.NavigateToChatInfo(cid)
-                    }
-                )
+                ChatClient.instance()
+                    .channel(cid)
+                    .truncate(Message(text = "System message"))
+                    .enqueue()
             }
         }
     }

```

</details>

### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist
